OliveDI/src/main/java/olive/di/DIContainer.kt

```diff
@@ -0,0 +1,147 @@
+        // abstact class 모듈인 경우 함수의 return 타입을 key로, 함수의 파라미터 타입을 value로 저장
+        // 파라미터 타입에 맞는 객체를 createSingleton로 직접 생성
+        val functions = this.declaredMemberFunctions
+        functions.forEach { function ->
+            val instanceType = function.parameters().first { it != this }
+            val cacheType = function.returnType.classifier as KClass<*>
+
+            if (function.annotations.isQualifierAnnotation()) { // Qualifier가 붙은 경우 namedInstances에 저장
+                val nameAnnotation = function.annotations.qualifierNameAnnotation()
+                namedInstances[cacheType] = NamedInstance(nameAnnotation, create(instanceType))
+                return@forEach
+            }
+            createSingleton(instanceType, cacheType)
+        }
+    }
+
+    private fun KFunction<*>.parameters(): List<KClass<*>> = parameters.map { it.type.classifier as KClass<*> }
+
+    private fun List<Annotation>.isQualifierAnnotation(): Boolean {
+        return any { annotation ->
+            annotation.annotationClass.annotations.any { it.annotationClass == Qualifier::class }
+        }
+    }
+
+    private fun List<Annotation>.qualifierNameAnnotation(): KClass<out Annotation> {
+        val nameAnnotation =
+            first { functionAnnotation ->
+                functionAnnotation.annotationClass.annotations.any { it.annotationClass == Qualifier::class }
+            }
+        return nameAnnotation.annotationClass
+    }
+
+    fun <T : Any> instance(classType: KClass<T>): T {
+        return create(classType)
+    }
+
+    fun <T : Any> singletonInstance(classType: KClass<T>): T {
+        return instances[classType] as? T ?: createSingleton(classType)
+    }
+
+    private fun <T : Any> createSingleton(
+        instanceType: KClass<T>,
+        cacheType: KClass<*> = instanceType,
+    ): T {
+        return create(instanceType).apply {
+            putSingletonInstance(cacheType)
+        }
+    }
+
+    private fun <T : Any> create(classType: KClass<T>): T {
+        if (instances[classType] != null) return instances[classType] as T
+        val parameters = classType.constructors.first().parameters
+        val arguments = parameters.map(::argumentInstance)
+        val instance = classType.constructors.first().call(*arguments.toTypedArray())
+        instance.injectFieldDependency()
+        return instance
+    }
+
+    private fun argumentInstance(parameter: KParameter): Any {
+        val parameterType = parameter.type.classifier as KClass<*>
+        if (parameter.annotations.isQualifierAnnotation()) {
+            return namedInstance(parameterType, parameter.annotations)
+        }
+        return instances[parameterType] ?: createSingleton(parameterType)
+    }
+
+    private fun Any.putSingletonInstance(type: KClass<*>) {
+        if (instances.containsKey(type)) {
+            throw IllegalArgumentException("이미 해당 클래스의 인스턴스가 존재합니다.")
+        }
+        instances[type] = this
```

Comment:
`instances.putIfAbsent(type,this)`
이렇게 쓸 수 도 있을 것 같아요!

Reply:
```suggestion
        instances.putIfAbsent(type, this) ?: throw IllegalArgumentException("이미 해당 클래스의 인스턴스가 존재합니다.")
```
그러면 위 4줄이 이렇게 한줄로 변경될 수 있겠네요~! `putIfAbsent` 함수에서 key 값이 없는 경우 `null`이 반환된다는 것이 명시적이지 않은 것 같아서, 일단 이대로 두겠습니당. ㅎㅎ
서기 덕분에 유용한 함수 알아가요 👍